src/sysmon/agent.py

```diff
@@ -0,0 +1,650 @@
+        retval = self._format_return(retval)
+        return retval
+
+    @RPC.export('network_io')
+    def network_io(self, per_nic=False, no_wrap=True, included_nics=None, sub_points=None):
+        """Return network input/output statistics."""
+        if included_nics and not per_nic:
+            per_nic = True
+        io_stats = psutil.net_io_counters(pernic=per_nic, nowrap=no_wrap)
+        retval = self._process_statistics(io_stats, sub_points, includes=included_nics)
+        self._get_throughput(io_stats, retval, per_nic, sub_points, 'receive_throughput', 'send_throughput',
+                             'bytes_recv', 'bytes_sent', self.last_network_received_bytes, self.last_network_sent_bytes)
+        retval = self._format_return(retval)
+        return retval
+
+    @RPC.export('network_connections')
+    def network_connections(self, kind='inet', sub_points=None):
+        """Return system-wide socket connections"""
+        connections = psutil.net_connections(kind)
+        connections = self._process_statistics(connections, sub_points=sub_points, format_return=False)
+        for k, v in connections.items():
+            if 'family' in v:
+                v['family'] = v['family'].name
+            if 'type' in v:
+                v['type'] = v['type'].name
+            if 'laddr' in v:
+                v['laddr'] = v['laddr'].ip + ':' + str(v['laddr'].port) \
+                    if type(v['laddr']) is psutil._common.addr else ''
+            if 'raddr' in v:
+                v['raddr'] = v['raddr'].ip + ':' + str(v['raddr'].port) \
+                    if type(v['raddr']) is psutil._common.addr else ''
+        connections = self._format_return(connections)
+        return connections
+
+    @RPC.export('network_interface_addresses')
+    def network_interface_addresses(self, included_interfaces=None, sub_points=None):
+        """Return addresses associated with network interfaces."""
+        addresses = psutil.net_if_addrs()
+        addresses = self._process_statistics(addresses, sub_points, includes=included_interfaces, format_return=False)
+        for k, v in addresses.items():
+            for item in v:
+                if 'family' in item:
+                    item['family'] = item['family'].name
+        addresses = self._format_return(addresses)
+        return addresses
+
+    @RPC.export('network_interface_statistics')
+    def network_interface_statistics(self, included_interfaces=None, sub_points=None):
+        """Return information about each network interface."""
+        stats = psutil.net_if_stats()
+        stats = self._process_statistics(stats, sub_points, includes=included_interfaces, format_return=False)
+        for k, v in stats.items():
+            if 'duplex' in v:
+                v['duplex'] = v['duplex'].name
+        stats = self._format_return(stats)
+        return stats
+
+    # TODO: Currently marked record only publish, as psutil bug #1708 duplicates core temps.
```

Comment:
Psutil bug 1708 was closed a really long time ago. Now, I'm not sure whether the TODO was telling me to work around the bug or to remove the workaround after it was fixed. We should confirm that this function is parsing its output correctly. The bug caused there to be duplicates (the names were duplicated, I think). So, when we call this function, does:
a) the return make sense?
b) the return match what we would get from psutil in ipython?
If we are losing information or failing to parse what comes back, we need to fix it. Othewise, we need to remove this TODO.

Reply:
Okay, I will test this week on a bare metal machine.